### ISSUE TITLE
Getting AtlasSession inside custom transformations

### ISSUE THREAD

**Issue by sanjay-90210** [atlasmap/atlasmap#2463]
I am looking at adding db based lookup for atlas maps. For a given document like purchase orders, customer item id may have to be mapped to the supplier item id when the document is transformed from a customer PO to a supplier PO, The list of items will be huge to allow it all to be loaded into system properties and in a multi tenant system, that is not even possible. 

I am looking at extending the default context and session to be able to associate one or more lookup objects with the session along with custom transforms to allow looking up values that are in the inbound document before they are added into the target document. There is no way to associate such an object into a session as of now, I guess and so the extended context and session objects. Even with this, the session is not getting passed into the transform functions so that I can get the requested lookup object to support this functionality. Is there anyway to get the session passed into the transform functions ?

I see an open issue [1269](https://github.com/atlasmap/atlasmap/issues/1269) that talks about injecting session into the field actions. I am looking at something like the one given below:

```
public class MyPurchaseOrderAction implements AtlasFieldAction
{
    @AtlasActionProcessor
    public static String poLookup(AtlasSession session, PurchaseOrderLookup action, String input)
    {
        // Get lookup object from session, call lookup with input and return the result
        LookupManager lm = (LookupManager)session.get(action.name);
        return lm.lookup(input);
    }
}
```

Is this the right approach or is there any existing mechanism for doing this ?

**Comment by igarashitm** [comment]
What I have in mind is like having annotated field in custom action class
```java
@AtlasSession
private AtlasSession session;
```
and yeah, would add some API to access source fields directly, possibly just delegate to `AtlasModule.readSourceValue()`, by specifying full field path expression
```
<Document ID>:/<Absolute field path>
```

**Comment by sanjay-90210** [comment]
Functions annotated with @AtlasActionProcessor needs to be static, I guess ? Injecting the AtlasSession value into the action class will mean that they will have to be non static with multiple action classes to be created for multiple sessions executing in parallel. 

If the annotation is on the function, as given below, and you pass the session as an argument, things can remain the same, except for the additional argument. Comments?

```
public class MyPurchaseOrderAction implements AtlasFieldAction
{
    @AtlasActionProcessor
    @AtlasSessionArgument
    public static String poLookup(AtlasSession session, PurchaseOrderLookup action, String input)
    {
        // Get lookup object from session, call lookup with input and return the result
        LookupManager lm = (LookupManager)session.get(action.name);
        return lm.lookup(input);
    }
}
```

**Comment by igarashitm** [comment]
IIRC it doesn't have to be static, but yeah, method arg injection would be also an option.

**Comment by sanjay-90210** [comment]
Ah, good to know. Either way, Getting access to the AtlasSession in a custom transform will solve my issue. Without that, I will have to setup a thread local variable to get the session, which I would want to avoid.

Is there a way to associate objects with an AtlasSession other than extending the factory and context classes to add a map to do this?

**Comment by pleacu** [comment]
Try this: https://docs.atlasmap.io/#creating-a-custom-transformation
  --Paul

**Comment by sanjay-90210** [comment]
Hi pleacu,
        I actually created and tested a custom transform and came across the issue state in the beginning.

BTW, The documentation that you point to was used in coding the initial version and it did not work as it did not say anything about the META-INF/services requirement. I had to dig that information out of one of your test cases. It will be nice if the services entry requirement is added to the above document.

**Comment by igarashitm** [comment]
If you're handling AtlasSession directly (i.e. not through camel-atlasmap), you might be able to leverage session property. See `AtlasSession#getSourceProperties()` and etc.

**Comment by sanjay-90210** [comment]
I saw the source and target properties in the session. I guess i can set a property without defining it in the map that does not collide with the user defined properties and sneak an object in ? I saw some types in DefaultAtlasPropertyStrategy and thought that it is validated somehow with no option to set the type from the ui and so ignored it.

private List<AtlasPropertyType> propertyOrder = Arrays.asList(
            AtlasPropertyType.RUNTIME_PROPERTIES,
            AtlasPropertyType.JAVA_SYSTEM_PROPERTIES,
            AtlasPropertyType.ENVIRONMENT_VARIABLES,
            AtlasPropertyType.MAPPING_DEFINED_PROPERTIES);

I will give it a try.

**Comment by sanjay-90210** [comment]
> ~If you're handling AtlasSession directly (i.e. not through camel-atlasmap)~ nevermind, custom action would be able to get/put them once AtlasSession injection is achieved, you might be able to leverage session property. See `AtlasSession#getSourceProperties()` and etc.

Yes, indeed. I tried it out and things worked perfectly. All I really need now is a way to get the session inside the transform function. I have setup a thread local mechanism as of now to verify.

```
public class DeviceAction implements AtlasFieldAction
{
    @AtlasActionProcessor
    public static String deviceLookup(DeviceLookup action, String input)
    {
        LookupManager lm = (LookupManager)SessionManager.getSession().
            getTargetProperties().get("com.example.lookup." + action.tableName);
        return lm.lookup(input);
    }
}
```
Is there a way to change the name of the transform as shown in the transformation selection ui ?

![transform](https://user-images.githubusercontent.com/73641844/98354167-5b825600-2046-11eb-9e64-e650b4aba774.png)

It shows the FQ java class name with the first letter capitalized. It will be nice to have a name like "example : Device Lookup" so that actions implemented in a deeply nested package structure does not show all that implementation package name in the ui. I tried overriding the getDisplayName in FieldAction, but that did not work.

```
public interface FieldAction {

    @JsonIgnore
    default String getDisplayName() {
        // TODO display name should be more human readable one instead of class name
        return this.getClass().getSimpleName();
    }
}
```

**Comment by igarashitm** [comment]
OK good to hear, right now I'm busy with others but I'll count your vote for #1269.
And yep, I think ATM providing display name through metadata is not working well. Can you file an issue?

**Comment by sanjay-90210** [comment]
Sorry, I got pulled into unrelated issues. I have filed an issue [2483](https://github.com/atlasmap/atlasmap/issues/2483) for the custom transform display issue. Please take a look.

**Comment by igarashitm** [comment]
Duplicate of #1269
